### DISCUSSION TITLE
Adding type hints

### DISCUSSION THREAD

**Discussion by dibyendumajumdar** [mjhoptics/ray-optics#61] [General]
Hi

I am interested in adding type hints to the code to make it easier to understand but as I try to do this, I see there are a few cycling dependencies between classes. Any thoughts on this? 

Regards
Dibyendu

**Comment by mjhoptics** [comment]
Yes, this is an issue. I spent many years working with C++ and python seemed to adopt type hints somewhat grudgingly. There are things you can use in C++ (forward referencing) to reduce the coupling between modules. Python doesn't seem to have equivalent features. 
Since type hints aren't used to enforce which types are used, I feel that adding to the documentation of the class or method or function in question is the better course. I'm happy to answer questions or add info for the documentation. I've noticed that the sphinx doc build is quite sensitive to circular references, and it can be baffling to try to figure them out. 
What do you think?
Mike

**Comment by dibyendumajumdar** [comment]
Y it is a difficult one. For readers of the code lack of type info makes it difficult to follow what is happening.
For this reason too, adding hints to existing code is not straight forward. I have to stop in debugger to see what the types are...

It seems if everything was annotated then mypy can be used to validate the code conforms to the types.

I started on adding hints, but I feel like I might be better off just porting to Java (there is a already a C++ port of your project). 

If I do manage to get somewhere with this, will open a pull request. It is not clear to me if its worth it in Python; the code gets more cluttered obviously

**Reply by mjhoptics** [reply]
Feel free to submit pull requests with what you're doing. There are probably some places where type hints would be the best way to document code. I'd be interested in seeing any examples you develop.

**Comment by dibyendumajumdar** [comment]
Re the circular dependencies, I think it is possible to work around them; I am trying an approach.

One other 'issue' I found is that  Python has no real named constants; so when variables are used as constants it breaks the type checker.  For example value[0] works but value[Intfc] doesn't with the type checker

**Reply by mjhoptics** [reply]
Yes, this is an area that I've struggled with in python. The use of constants such as `Intfc` was an attempt to use a C++ style I was comfortable with python. Hard coding the numeric indices like value[0] really bothers me (from long experience with other languages like Fortran).
When I write code now where a tuple is an argument or return value, rather than use named constants, I'll convert to/from the tuple and individual variables at the place of use. I also try using similar variable names at both ends of the conversion. The analyses module in the raytr package has a lot of this style of usage. e.g.
```
    image_pt, ref_dir, ref_sphere_radius = ref_sphere
    cr, cr_exp_seg = chief_ray_pkg
    chief_ray, chief_ray_op, wvl = cr
    cr_exp_pt, cr_exp_dir, cr_exp_dist, ifc, cr_b4_pt, cr_b4_dir = cr_exp_seg
```
If a variable has a suffix `_pkg` then it is a tuple that exists solely to package together some related data items.

I've also moved away from using python enumerated types in favor of simple strings. These are as clear in the code as enums (the way I use them, at least), fail similarly if you mistype the value, and stream without any special processing.

Let me know how your investigation with circular dependencies turns out.

Mike

**Comment by dibyendumajumdar** [comment]
Re circular dependencies I am trying three things:

1. In type annotations, we can enclose the type in quotes when we need forward references
2. We can use a super type with some signatures when the dependent code doesn't need implementation details - e.g. a protocol 
3. In some cases we can pass arguments to method and avoid a dependency . e.g. FieldSpec depends on FirstOrderData but this can be avoided by just passing required values in methods; also makes the methods more reusable

For what its worth you can look at what I am trying here: https://github.com/dibyendumajumdar/ray-optics/tree/typing

I just started 2 days ago so very early days

**Reply by mjhoptics** [reply]
Hi Dibyendu,
I've looked at (some) of your changes and have a few comments.
1) They really **are** type *hints*. The hint can be "wrong" and it doesn't matter. As a C++ programmer I find this confusing. As a Python programmer, I guess it's "what did you expect?"
The example in particular is:
Transform3 = Tuple[Vector3, Vector3]
It really should be:
Transform3 = Tuple[Matrix3, Vector3]
but since Matrix3 and Vector3 (and Vector2) all are defined as np.array, it doesn't have any info to verify this, and in fact it probably couldn't until run time. So this is less than ideal.
2) I'm not sure it's good practice to reuse `typing` as a module name, hiding the system one. I can see why you did it (fewer imports) but I'd want to consider using, e.g. `model_typing`.
3) There's a tradeoff between having the type definitions in a separate file (which reduces circular dependencies) vs having the type definitions where the types are originally defined. This seems like an opportunity for the two to get out of sync. Since the type definitions are really only hints, there's a chance that the code could be incorrect but not flagged by the interpreter. Also, the doc build has no way (currently) to tie these two together.
4) Be careful about changing signatures. I have a couple of notebooks that call `obj_coords`, and it's conceivable that there are other examples in the field. By the way, I disagree with the change to obj_coords. The intent of a call with a Field argument was to push all of the responsibility for the calculation inside of obj_coords. Field has the necessary relationship with OpticalSpec to get the info needed for the calculation. Having the calling site responsible for passing supporting data for the calculation breaks this encapsulation.
5) I like the idea of Literal definitions for the various string constants that are used in various places. 
6) Probably the biggest shortcoming of ray-optics is lack of test coverage. There are a few unit tests that can be run and I generally run a subset of the notebooks to ensure they still work. Making large changes to a code base without commensurate test coverage makes me nervous.

7) I think there is a larger, philosophical, question for all of this. That is, what is the appropriate style to use for a medium scale python program? As I've written more python and my C++ programming days get further into the past, I've been embracing using python basic types for more uses (e.g. strings instead of enums, dicts instead of simple classes), writing code assuming duck typing rather than formal inheritance hierarchies, more use of closures, etc. Part of this is a judgement on my part of trying to use the language pythonically. I've seen a lot of examples over time of programmers trying to use one language like another one (e.g. FORTRAN programmers trying to write FORTRAN using C) and I'm consciously trying to avoid writing C++ using python.

I think what you're doing is valuable but also risky, because of lack of test coverage. Make sure saving and restoring lenses works. You can use the notebooks OpenModelsTest and Sasian Triplet to do bare bones smoke tests on your changes.

**Reply by dibyendumajumdar** [reply]
> Hi Dibyendu, I've looked at (some) of your changes and have a few comments.

Hi Mike, thank you for detailed review - I appreciate it!

>     1. They really **are** type _hints_. The hint can be "wrong" and it doesn't matter. As a C++ programmer I find this confusing. As a Python programmer, I guess it's "what did you expect?"

It is true that for the Python interpreter these hints are no-op, but a type checker can catch inconsistency. 

>        The example in particular is:
>        Transform3 = Tuple[Vector3, Vector3]
>        It really should be:
>        Transform3 = Tuple[Matrix3, Vector3]
>        but since Matrix3 and Vector3 (and Vector2) all are defined as np.array, it doesn't have any info to verify this, and in fact it probably couldn't until run time. So this is less than ideal.

The issue with ndarray specifically is that we can't distinguish between Matrix and Vector - they all look the same. I thought of using aliases that at least help improve the understanding the code.

Btw I had already spotted the issue with Transform3 and fixed it; but maybe you read the earlier commit. Note that to allow None as the first element I had to mark it as Optional.

```
Transform3 = Tuple[Optional[Matrix3], Vector3]
```


> 
>     2. I'm not sure it's good practice to reuse `typing` as a module name, hiding the system one. I can see why you did it (fewer imports) but I'd want to consider using, e.g. `model_typing`.

Yes it wasn't clear to me what best practice is; I saw other projects also using the typing module so I thought maybe that is best practice. But can be changed no issue with that.

> 
>     3. There's a tradeoff between having the type definitions in a separate file (which reduces circular dependencies) vs having the type definitions where the types are originally defined. This seems like an opportunity for the two to get out of sync. Since the type definitions are really only hints, there's a chance that the code could be incorrect but not flagged by the interpreter. Also, the doc build has no way (currently) to tie these two together.

True, and eventually the types should be moved to the right places. Although Python doesn't look at these type hints, mypy will catch inconsistencies if properly annotated subject to the issue with ndarray of course.


> 
>     4. Be careful about changing signatures. I have a couple of notebooks that call `obj_coords`, and it's conceivable that there are other examples in the field. By the way, I disagree with the change to obj_coords. The intent of a call with a Field argument was to push all of the responsibility for the calculation inside of obj_coords. Field has the necessary relationship with OpticalSpec to get the info needed for the calculation. Having the calling site responsible for passing supporting data for the calculation breaks this encapsulation.
> 

Okay. I don't yet understand the dependencies, and it seemed that there was unnecessary coupling there. However you are right that it is better to not change any signatures. I wasn't planning to. I will revert this change. 

>     5. I like the idea of Literal definitions for the various string constants that are used in various places.
> 
>     6. Probably the biggest shortcoming of ray-optics is lack of test coverage. There are a few unit tests that can be run and I generally run a subset of the notebooks to ensure they still work. Making large changes to a code base without commensurate test coverage makes me nervous.
> 

Yes agreed. I wonder if some of your example notebooks can be converted to tests; they won't be unit tests; more like integration tests; but useful still to catch issues. 

>     7. I think there is a larger, philosophical, question for all of this. That is, what is the appropriate style to use for a medium scale python program? As I've written more python and my C++ programming days get further into the past, I've been embracing using python basic types for more uses (e.g. strings instead of enums, dicts instead of simple classes), writing code assuming duck typing rather than formal inheritance hierarchies, more use of closures, etc. Part of this is a judgement on my part of trying to use the language pythonically. I've seen a lot of examples over time of programmers trying to use one language like another one (e.g. FORTRAN programmers trying to write FORTRAN using C) and I'm consciously trying to avoid writing C++ using python.

Indeed - I am not sure either whether it is helpful to add type hints to Python code or not.
I think once I start to understand the code, it will probably become less important to me personally whether type annotations are there or not. The main advantages are:

* Type checker like mypy can catch type errors
* It makes it easier for new comers to grok the code
* IDEs can benefit a lot - I use PyCharm and it immediately reacts to type hints.

> 
> I think what you're doing is valuable but also risky, because of lack of test coverage. Make sure saving and restoring lenses works. You can use the notebooks OpenModelsTest and Sasian Triplet to do bare bones smoke tests on your changes.

Yes good point. I was not really planning to make changes that would break anything - so I will go back and revisit the breaking change anyway.

**Comment by dibyendumajumdar** [comment]
Mike, thank you for the all the feedback on my changes. 
I will continue working on it until I can get coverage of all the 'backend' code. The  front-end or UI code doesn't excite me so much, so I will probably skip those.

If at some point you think it is worth it then I can open a PR.

**Reply by mjhoptics** [reply]
The work you've done so far is valuable and I think it should be moved into the main dev branch soon. I've been reviewing your repo so have been able to keep up with the changes you're making. I think I should do a point release of ray-optics and then you can open a PR. This gives a stable fall back release point. Rather than build up a month's worth of changes, it's better to do more frequent integrations, I think.
Thanks for your work on this.  Mike

**Reply by dibyendumajumdar** [reply]
I also need to find a solution of the named constants as mypy doesn't like those.

Btw I was wondering if the naming convention for those constants can be changed - like making them upper case or prefixing them with `k`, for example. It is hard to tell them apart at the moment.

**Reply by mjhoptics** [reply]
I've tried 2 things with named constants:
1) import model constants as `mc` and then refer to the named constants as `mc.Gap`, `mc.slp`, `mc.ht`, etc.
2) Not use them at all in favor of tuple expansion, e.g.
`ray, opd, wl = ray_pkg`
`ifc, g, tfrm_after, rindx_after, z_dir_after = after`

I think #2 is actually "more pythonic" and I've been writing more new code in this style. Where #2 isn't feasible for some reason (?), I'd prefer choice #1 instead of importing the named constants directly and using them without qualification. But I really don't like referring to tuple items by their numeric index; this seems both brittle and unclear to the reader of the code.

**Reply by dibyendumajumdar** [reply]
I have pushed a commit that contains a solution that seems to help type checking work when named constants are used in place of literals
